rita/src/rita_common/dao_manager/mod.rs

```diff
@@ -0,0 +1,257 @@
+                on_list: response.result == Uint256::zero(),
+                dao_address: dao_address,
+                id: their_id.clone(),
+                last_updated: Instant::now(),
+            };
+            let mut v = Vec::new();
+            v.push(entry);
+            self.entries.insert(their_id, v);
+            // send appropriate dao message
+        }
+    }
+}
+
+/// True if timestamp does not need to be updated
+fn timer_check(timestamp: Instant) -> bool {
+    if Instant::now() - timestamp < SETTING.get_dao().cache_timeout {
+        true
+    } else {
+        false
+    }
+}
+
+/// Checks is an identity is in at least one of the set of DAO's we are a member of.
+/// will check the cache first before going out and updating via web3
+fn check_cache(their_id: Identity, entries: &HashMap<Identity, Vec<DAOEntry>>) -> () {
+    // we don't care about subnet DAO's, short circuit.
+    if !SETTING.get_dao().dao_enforcement {
+        trace!("DAO enforcement disabled DAOMAnager doing nothing!");
+        return ();
+    }
+    // TODO when we start enforcing dao state more strictly we will need
+    // to detect when we are bootstrapping and explicitly allow everyone
+
+    match entries.get(&their_id) {
+        // Cache hit
+        Some(membership_list) => {
+            for entry in membership_list.iter() {
+                if entry.on_list && timer_check(entry.last_updated) {
+                    trace!(
+                        "{:?} is on the SubnetDAO {:?}",
+                        their_id.clone(),
+                        entry.dao_address
+                    );
+                // send approval
+                } else if !timer_check(entry.last_updated) {
+                    get_membership(entry.dao_address, entry.id.clone());
+                }
+            }
+            trace!("{:?} is not on any SubnetDAO", their_id);
+            // send rejection
+        }
+        // Cache miss, do a lookup for all DAO's
+        None => {
+            for dao in SETTING.get_dao().dao_addresses.iter() {
+                get_membership(dao.clone(), their_id.clone());
+            }
+        }
+    }
+}
+
+fn get_membership(dao_address: EthAddress, target: Identity) -> () {
+    let (url, port) = get_web3_server();
+    let endpoint = format!("{}:{}/", url, port);
+    let socket: SocketAddr = endpoint.parse().expect("Invalid DAO fullnode!");
+
+    // We transform the ip address into a argument
+    let ip_bytes = match target.mesh_ip {
+        IpAddr::V6(ip) => ip.octets(),
+        _ => panic!("MeshIP must be ipv6 and is not!"),
+    };
+    let mut full_bytes: [u8; 32] = [0; 32];
+    let mut i = 0;
+    for byte in ip_bytes.iter() {
+        full_bytes[i] = byte.clone();
+        i = i + 1;
+    }
+
+    let call_args: Uint256 = full_bytes.into();
+    let func_call = format!("{{'jsonrpc':'2.0','method':'eth_call','params':[{{'to': '{:x}', 'data': '{:x}'}}, 'latest'],'id':1}}", dao_address, call_args);
```

Comment:
Just for brevity we could serialize proper struct that imitates JSONRPC request.

Reply:
This doesn't feel right to me either.

Reply:
the real issue is the nesting required, the struct gets rather complicated looking. Not to mention how do I get 'latest' to serialize without a field name? If you can give me a struct that serializes this way I'll be happy but I don't know how to make one.